scipy/optimize/_minpack_py.py

```diff
@@ -797,6 +810,22 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
     if ydata.size == 0:
         raise ValueError("`ydata` must not be empty!")
 
+    # nan handling is needed only if check_finite is False because if True,
+    # the x-y data are already checked, and they don't contain nans.
+    if not check_finite:
+        x_contains_nan, nan_policy = _contains_nan(xdata, nan_policy)
+        y_contains_nan, nan_policy = _contains_nan(ydata, nan_policy)
+
+        if (x_contains_nan or y_contains_nan) and nan_policy == 'omit':
+            # ignore NaNs for N dimensional arrays
+            has_nan = np.isnan(xdata)
+            for _ in range(xdata.ndim-1):
+                has_nan = has_nan.any(axis=0)
+            has_nan |= np.isnan(ydata)
+
+            xdata = xdata[..., ~has_nan]
+            ydata = ydata[~has_nan]
```

Comment:
Based on the doc:
https://github.com/scipy/scipy/blob/e2fffa9bc3a872f8c30da6f675b05b8c30f92c94/scipy/optimize/_minpack_py.py#L548-L552
`xdata` should be 1D or 2D array, but I tried a 3D data, it works.

```
In [38]: xdata = np.array([[[2, 3, 4, 4,],
    ...:                    [2, 3, 4, 4]]]
    ...:                    )

In [39]: ydata = np.array([1,2,5,3])

In [40]: def f(x, a, b):
    ...:     x1 = x[0, 0, :]
    ...:     x2 = x[0, 1, :]
    ...:     return a*x1+b+x2

In [42]: optimize.curve_fit(f, xdata, ydata)
Out[42]:
(array([ 0.54545455, -2.27272727]),
 array([[ 0.39669421, -1.28925619],
        [-1.28925619,  4.46280986]])) 
```

So, these codes work N dimensional xdata.

Reply:
Really? Looks to me like they will produce ragged arrays of there are different numbers of NaNs in different slices.

Reply:
I suppose we should support Nd since this used to be
![image](https://user-images.githubusercontent.com/6570539/208184448-2b35552e-721a-40fa-8845-7fd78a8bc660.png)

But what did you think of using an `axis` tuple instead of looping? (see comment above)

Reply:
> Why would inputs be 3D? What would that mean here - that the function accepts matrix inputs?
The documentation said it would only accept 1D or 2D data, but out of curiosity I entered 3D data and it returned a value.
So, if some users had already written code with 3D data input, I thought it would be better not to break their codes.
If we can ignore the current behavior, I will remove the code for it. What do you think?
I used your tuple idea now.